storage/lmdb/src/remap.rs

```diff
@@ -0,0 +1,135 @@
+    pub const fn from_bytes(bytes: usize) -> Self {
+        Self(bytes)
+    }
+
+    /// Specify in the number of kilobytes
+    pub const fn from_kilobytes(kilobytes: usize) -> Self {
+        Self::from_bytes(1024 * kilobytes)
+    }
+
+    /// Specify in the number of megabytes
+    pub const fn from_megabytes(megabytes: usize) -> Self {
+        Self::from_kilobytes(1024 * megabytes)
+    }
+
+    /// Convert to raw byte count
+    pub const fn as_bytes(self) -> usize {
+        self.0
+    }
+
+    pub fn div_ceil(self, rhs: Self) -> usize {
+        self.0 / rhs.0 + (self.0 % rhs.0 > 0) as usize
+    }
+}
+
+impl std::fmt::Display for MemSize {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}B", self.0)
+    }
+}
+
+impl std::ops::Add for MemSize {
+    type Output = MemSize;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Self(self.0 + rhs.0)
+    }
+}
+
+impl std::ops::Mul<MemSize> for usize {
+    type Output = MemSize;
+
+    fn mul(self, rhs: MemSize) -> Self::Output {
+        MemSize(self * rhs.0)
+    }
+}
+
+/// Token representing acquisition of the memory map resource
+#[derive(Debug)]
+pub struct MemMapToken();
+
+impl MemMapToken {
+    pub fn new() -> Self {
+        Self()
+    }
+}
+
+/// A proof of having acquired the memory map resource exclusively
+pub type ExclusiveMemMapToken<'a> = std::sync::RwLockWriteGuard<'a, MemMapToken>;
+
+/// Memory remapping procedure. Ensure at least `headroom` free space is available
+pub fn remap(
+    env: &lmdb::Environment,
+    _map_token: ExclusiveMemMapToken<'_>,
+    headroom: MemSize,
+) -> storage_core::Result<()> {
+    // Get page size
+    let page_size = {
+        let stat = env.stat().or_else(crate::error::process_with_err)?;
+        MemSize::from_bytes(stat.page_size() as usize)
+    };
+
+    // Get current occupancy info
+    let info = env.info().or_else(crate::error::process_with_err)?;
+    let current_size = MemSize::from_bytes(info.map_size());
+    let current_pages = current_size.div_ceil(page_size);
+    let used_pages = info.last_pgno() + 1;
+    let freelist_pages = env.freelist().or_else(crate::error::process_with_err)?;
+    let free_pages = (current_pages - used_pages) + freelist_pages;
+
+    // Get map size requirements
+    let required_free_pages = headroom.div_ceil(page_size);
+    let required_pages = used_pages + required_free_pages;
+    let required_size = required_pages * page_size;
+
+    log::trace!(
+        "LMDB memory: occupied = {} - {}, reserved = {}, free = {}, required = {}",
+        used_pages * page_size,
+        freelist_pages * page_size,
+        current_size,
+        free_pages * page_size,
+        required_size,
+    );
+
+    // Ensure there is enough space
+    if required_size > current_size {
+        // Remap with double the size but or the required size, whichever is larger
+        let new_size = std::cmp::max(2 * current_size, required_size);
```

Comment:
Could you please explain this logic to me? Why in one case (`required_size < 2 * current_size`) we want to have an additional capacity, but in other case (`required_size > 2 * current_size`) we don't?

Reply:
The basic idea is to double the capacity if we run out. However, in some rare cases (when the initial size is small and the expected current tx size is large), that would not be sufficient. The `max` is there to remain correct even in these rare cases.

Reply:
I got that, but would it make sense to double `required_size` too?..

Reply:
> I got that, but would it make sense to double `required_size` too?..

Yeah why not, will do